<!DOCTYPE html>
<!-- saved from url=(0130)file://localhost/private/var/folders/32/z8pttxw57h57tgpd2pfljgtw0000gn/T/RtmpyGoYkc/preview-e73f6b584bba.dir/ActiveDriverWGSR.html -->
<html xmlns="http://www.w3.org/1999/xhtml"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8">



<meta name="generator" content="pandoc">

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Helen Zhu, Dr.Â Juri Reimand">

<meta name="date" content="2019-01-20">

<title>ActiveDriverWGSR</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css">

<script type="text/javascript" src="./ActiveDriverWGSR_files/MathJax.js"></script><style type="text/css">.MathJax_Hover_Frame {border-radius: .25em; -webkit-border-radius: .25em; -moz-border-radius: .25em; -khtml-border-radius: .25em; box-shadow: 0px 0px 15px #83A; -webkit-box-shadow: 0px 0px 15px #83A; -moz-box-shadow: 0px 0px 15px #83A; -khtml-box-shadow: 0px 0px 15px #83A; border: 1px solid #A6D ! important; display: inline-block; position: absolute}
.MathJax_Menu_Button .MathJax_Hover_Arrow {position: absolute; cursor: pointer; display: inline-block; border: 2px solid #AAA; border-radius: 4px; -webkit-border-radius: 4px; -moz-border-radius: 4px; -khtml-border-radius: 4px; font-family: 'Courier New',Courier; font-size: 9px; color: #F0F0F0}
.MathJax_Menu_Button .MathJax_Hover_Arrow span {display: block; background-color: #AAA; border: 1px solid; border-radius: 3px; line-height: 0; padding: 4px}
.MathJax_Hover_Arrow:hover {color: white!important; border: 2px solid #CCC!important}
.MathJax_Hover_Arrow:hover span {background-color: #CCC!important}
</style><style type="text/css">#MathJax_About {position: fixed; left: 50%; width: auto; text-align: center; border: 3px outset; padding: 1em 2em; background-color: #DDDDDD; color: black; cursor: default; font-family: message-box; font-size: 120%; font-style: normal; text-indent: 0; text-transform: none; line-height: normal; letter-spacing: normal; word-spacing: normal; word-wrap: normal; white-space: nowrap; float: none; z-index: 201; border-radius: 15px; -webkit-border-radius: 15px; -moz-border-radius: 15px; -khtml-border-radius: 15px; box-shadow: 0px 10px 20px #808080; -webkit-box-shadow: 0px 10px 20px #808080; -moz-box-shadow: 0px 10px 20px #808080; -khtml-box-shadow: 0px 10px 20px #808080; filter: progid:DXImageTransform.Microsoft.dropshadow(OffX=2, OffY=2, Color='gray', Positive='true')}
#MathJax_About.MathJax_MousePost {outline: none}
.MathJax_Menu {position: absolute; background-color: white; color: black; width: auto; padding: 5px 0px; border: 1px solid #CCCCCC; margin: 0; cursor: default; font: menu; text-align: left; text-indent: 0; text-transform: none; line-height: normal; letter-spacing: normal; word-spacing: normal; word-wrap: normal; white-space: nowrap; float: none; z-index: 201; border-radius: 5px; -webkit-border-radius: 5px; -moz-border-radius: 5px; -khtml-border-radius: 5px; box-shadow: 0px 10px 20px #808080; -webkit-box-shadow: 0px 10px 20px #808080; -moz-box-shadow: 0px 10px 20px #808080; -khtml-box-shadow: 0px 10px 20px #808080; filter: progid:DXImageTransform.Microsoft.dropshadow(OffX=2, OffY=2, Color='gray', Positive='true')}
.MathJax_MenuItem {padding: 1px 2em; background: transparent}
.MathJax_MenuArrow {position: absolute; right: .5em; padding-top: .25em; color: #666666; font-size: .75em}
.MathJax_MenuActive .MathJax_MenuArrow {color: white}
.MathJax_MenuArrow.RTL {left: .5em; right: auto}
.MathJax_MenuCheck {position: absolute; left: .7em}
.MathJax_MenuCheck.RTL {right: .7em; left: auto}
.MathJax_MenuRadioCheck {position: absolute; left: .7em}
.MathJax_MenuRadioCheck.RTL {right: .7em; left: auto}
.MathJax_MenuLabel {padding: 1px 2em 3px 1.33em; font-style: italic}
.MathJax_MenuRule {border-top: 1px solid #DDDDDD; margin: 4px 3px}
.MathJax_MenuDisabled {color: GrayText}
.MathJax_MenuActive {background-color: #606872; color: white}
.MathJax_MenuDisabled:focus, .MathJax_MenuLabel:focus {background-color: #E8E8E8}
.MathJax_ContextMenu:focus {outline: none}
.MathJax_ContextMenu .MathJax_MenuItem:focus {outline: none}
#MathJax_AboutClose {top: .2em; right: .2em}
.MathJax_Menu .MathJax_MenuClose {top: -10px; left: -10px}
.MathJax_MenuClose {position: absolute; cursor: pointer; display: inline-block; border: 2px solid #AAA; border-radius: 18px; -webkit-border-radius: 18px; -moz-border-radius: 18px; -khtml-border-radius: 18px; font-family: 'Courier New',Courier; font-size: 24px; color: #F0F0F0}
.MathJax_MenuClose span {display: block; background-color: #AAA; border: 1.5px solid; border-radius: 18px; -webkit-border-radius: 18px; -moz-border-radius: 18px; -khtml-border-radius: 18px; line-height: 0; padding: 8px 0 6px}
.MathJax_MenuClose:hover {color: white!important; border: 2px solid #CCC!important}
.MathJax_MenuClose:hover span {background-color: #CCC!important}
.MathJax_MenuClose:hover:focus {outline: none}
</style><style type="text/css">.MathJax_Preview .MJXf-math {color: inherit!important}
</style><style type="text/css">.MJX_Assistive_MathML {position: absolute!important; top: 0; left: 0; clip: rect(1px, 1px, 1px, 1px); padding: 1px 0 0 0!important; border: 0!important; height: 1px!important; width: 1px!important; overflow: hidden!important; display: block!important; -webkit-touch-callout: none; -webkit-user-select: none; -khtml-user-select: none; -moz-user-select: none; -ms-user-select: none; user-select: none}
.MJX_Assistive_MathML.MJX_Assistive_MathML_Block {width: 100%!important}
</style><style type="text/css">#MathJax_Zoom {position: absolute; background-color: #F0F0F0; overflow: auto; display: block; z-index: 301; padding: .5em; border: 1px solid black; margin: 0; font-weight: normal; font-style: normal; text-align: left; text-indent: 0; text-transform: none; line-height: normal; letter-spacing: normal; word-spacing: normal; word-wrap: normal; white-space: nowrap; float: none; -webkit-box-sizing: content-box; -moz-box-sizing: content-box; box-sizing: content-box; box-shadow: 5px 5px 15px #AAAAAA; -webkit-box-shadow: 5px 5px 15px #AAAAAA; -moz-box-shadow: 5px 5px 15px #AAAAAA; -khtml-box-shadow: 5px 5px 15px #AAAAAA; filter: progid:DXImageTransform.Microsoft.dropshadow(OffX=2, OffY=2, Color='gray', Positive='true')}
#MathJax_ZoomOverlay {position: absolute; left: 0; top: 0; z-index: 300; display: inline-block; width: 100%; height: 100%; border: 0; padding: 0; margin: 0; background-color: white; opacity: 0; filter: alpha(opacity=0)}
#MathJax_ZoomFrame {position: relative; display: inline-block; height: 0; width: 0}
#MathJax_ZoomEventTrap {position: absolute; left: 0; top: 0; z-index: 302; display: inline-block; border: 0; padding: 0; margin: 0; background-color: white; opacity: 0; filter: alpha(opacity=0)}
</style><style type="text/css">.MathJax_Preview {color: #888}
#MathJax_Message {position: fixed; left: 1em; bottom: 1.5em; background-color: #E6E6E6; border: 1px solid #959595; margin: 0px; padding: 2px 8px; z-index: 102; color: black; font-size: 80%; width: auto; white-space: nowrap}
#MathJax_MSIE_Frame {position: absolute; top: 0; left: 0; width: 0px; z-index: 101; border: 0px; margin: 0px; padding: 0px}
.MathJax_Error {color: #CC0000; font-style: italic}
</style><style type="text/css">.MJXp-script {font-size: .8em}
.MJXp-right {-webkit-transform-origin: right; -moz-transform-origin: right; -ms-transform-origin: right; -o-transform-origin: right; transform-origin: right}
.MJXp-bold {font-weight: bold}
.MJXp-italic {font-style: italic}
.MJXp-scr {font-family: MathJax_Script,'Times New Roman',Times,STIXGeneral,serif}
.MJXp-frak {font-family: MathJax_Fraktur,'Times New Roman',Times,STIXGeneral,serif}
.MJXp-sf {font-family: MathJax_SansSerif,'Times New Roman',Times,STIXGeneral,serif}
.MJXp-cal {font-family: MathJax_Caligraphic,'Times New Roman',Times,STIXGeneral,serif}
.MJXp-mono {font-family: MathJax_Typewriter,'Times New Roman',Times,STIXGeneral,serif}
.MJXp-largeop {font-size: 150%}
.MJXp-largeop.MJXp-int {vertical-align: -.2em}
.MJXp-math {display: inline-block; line-height: 1.2; text-indent: 0; font-family: 'Times New Roman',Times,STIXGeneral,serif; white-space: nowrap; border-collapse: collapse}
.MJXp-display {display: block; text-align: center; margin: 1em 0}
.MJXp-math span {display: inline-block}
.MJXp-box {display: block!important; text-align: center}
.MJXp-box:after {content: " "}
.MJXp-rule {display: block!important; margin-top: .1em}
.MJXp-char {display: block!important}
.MJXp-mo {margin: 0 .15em}
.MJXp-mfrac {margin: 0 .125em; vertical-align: .25em}
.MJXp-denom {display: inline-table!important; width: 100%}
.MJXp-denom > * {display: table-row!important}
.MJXp-surd {vertical-align: top}
.MJXp-surd > * {display: block!important}
.MJXp-script-box > *  {display: table!important; height: 50%}
.MJXp-script-box > * > * {display: table-cell!important; vertical-align: top}
.MJXp-script-box > *:last-child > * {vertical-align: bottom}
.MJXp-script-box > * > * > * {display: block!important}
.MJXp-mphantom {visibility: hidden}
.MJXp-munderover {display: inline-table!important}
.MJXp-over {display: inline-block!important; text-align: center}
.MJXp-over > * {display: block!important}
.MJXp-munderover > * {display: table-row!important}
.MJXp-mtable {vertical-align: .25em; margin: 0 .125em}
.MJXp-mtable > * {display: inline-table!important; vertical-align: middle}
.MJXp-mtr {display: table-row!important}
.MJXp-mtd {display: table-cell!important; text-align: center; padding: .5em 0 0 .5em}
.MJXp-mtr > .MJXp-mtd:first-child {padding-left: 0}
.MJXp-mtr:first-child > .MJXp-mtd {padding-top: 0}
.MJXp-mlabeledtr {display: table-row!important}
.MJXp-mlabeledtr > .MJXp-mtd:first-child {padding-left: 0}
.MJXp-mlabeledtr:first-child > .MJXp-mtd {padding-top: 0}
.MJXp-merror {background-color: #FFFF88; color: #CC0000; border: 1px solid #CC0000; padding: 1px 3px; font-style: normal; font-size: 90%}
.MJXp-scale0 {-webkit-transform: scaleX(.0); -moz-transform: scaleX(.0); -ms-transform: scaleX(.0); -o-transform: scaleX(.0); transform: scaleX(.0)}
.MJXp-scale1 {-webkit-transform: scaleX(.1); -moz-transform: scaleX(.1); -ms-transform: scaleX(.1); -o-transform: scaleX(.1); transform: scaleX(.1)}
.MJXp-scale2 {-webkit-transform: scaleX(.2); -moz-transform: scaleX(.2); -ms-transform: scaleX(.2); -o-transform: scaleX(.2); transform: scaleX(.2)}
.MJXp-scale3 {-webkit-transform: scaleX(.3); -moz-transform: scaleX(.3); -ms-transform: scaleX(.3); -o-transform: scaleX(.3); transform: scaleX(.3)}
.MJXp-scale4 {-webkit-transform: scaleX(.4); -moz-transform: scaleX(.4); -ms-transform: scaleX(.4); -o-transform: scaleX(.4); transform: scaleX(.4)}
.MJXp-scale5 {-webkit-transform: scaleX(.5); -moz-transform: scaleX(.5); -ms-transform: scaleX(.5); -o-transform: scaleX(.5); transform: scaleX(.5)}
.MJXp-scale6 {-webkit-transform: scaleX(.6); -moz-transform: scaleX(.6); -ms-transform: scaleX(.6); -o-transform: scaleX(.6); transform: scaleX(.6)}
.MJXp-scale7 {-webkit-transform: scaleX(.7); -moz-transform: scaleX(.7); -ms-transform: scaleX(.7); -o-transform: scaleX(.7); transform: scaleX(.7)}
.MJXp-scale8 {-webkit-transform: scaleX(.8); -moz-transform: scaleX(.8); -ms-transform: scaleX(.8); -o-transform: scaleX(.8); transform: scaleX(.8)}
.MJXp-scale9 {-webkit-transform: scaleX(.9); -moz-transform: scaleX(.9); -ms-transform: scaleX(.9); -o-transform: scaleX(.9); transform: scaleX(.9)}
.MathJax_PHTML .noError {vertical-align: ; font-size: 90%; text-align: left; color: black; padding: 1px 3px; border: 1px solid}
</style></head>

<body><div id="MathJax_Message" style="display: none;"></div>




<h1 class="title toc-ignore">ActiveDriverWGSR</h1>
<h4 class="author"><em>Helen Zhu, Dr.&nbsp;Juri Reimand</em></h4>
<h4 class="date"><em>2019-01-20</em></h4>



<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>ActiveDriverWGS is a driver discovery tool for cancer whole genome sequencing data. It analyzes the mutational burden of SNVs and short Indels in functionally defined regions of interest and returns regions which are significantly mutated compared to a background window. For more information, please refer to the ActiveDriverWGS preprint. <a href="https://www.biorxiv.org/content/early/2017/12/19/236802" class="uri">https://www.biorxiv.org/content/early/2017/12/19/236802</a></p>
<div id="please-note" class="section level3">
<h3>Please Note</h3>
<p>The genome build for ActiveDriverWGS is hg19. We are working on adding additional options for hg38 and mm10.</p>
</div>
</div>
<div id="input-data" class="section level1">
<h1>Input Data</h1>
<p>ActiveDriverWGSR requires a file for somatic mutations and a file for genomic elements. A third optional file for sites within genomic elements (i.e.&nbsp;post-translational modification sites in protein coding genes) can be used to find genomic elements with enriched site mutations.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ActiveDriverWGSR)

<span class="kw">data</span>(<span class="st">"cll_mutations"</span>)
<span class="kw">head</span>(cll_mutations)</code></pre></div>
<pre><code>##          chr      pos1      pos2 ref alt       patient  cc
## 779701  chr6  96651182  96651183  AA   T 001-0002-03TD CLL
## 779702 chr10 106556005 106556005   C   T 001-0002-03TD CLL
## 779703 chr10 107457352 107457352   C   T 001-0002-03TD CLL
## 779704 chr10 108111334 108111334   C   A 001-0002-03TD CLL
## 779705 chr10 109605024 109605024   T   A 001-0002-03TD CLL
##  [ reached getOption("max.print") -- omitted 1 row ]</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(<span class="st">"cancer_genes"</span>)
<span class="kw">head</span>(cancer_genes)</code></pre></div>
<pre><code>##      chr   start     end       id
## 648 chr1 2488103 2488172 TNFRSF14
## 649 chr1 2489164 2489273 TNFRSF14
## 650 chr1 2489781 2489907 TNFRSF14
## 651 chr1 2491261 2491417 TNFRSF14
## 652 chr1 2492062 2492153 TNFRSF14
## 653 chr1 2492932 2492963 TNFRSF14</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(<span class="st">"cancer_gene_sites"</span>)
<span class="kw">head</span>(cancer_gene_sites)</code></pre></div>
<pre><code>##         chr    start      end                  id
## 337831 chr1 11169412 11169412 MTOR NM_004958 2488
## 337834 chr1 11169413 11169413 MTOR NM_004958 2488
## 337837 chr1 11169414 11169414 MTOR NM_004958 2487
## 337839 chr1 11169415 11169415 MTOR NM_004958 2487
## 337841 chr1 11169416 11169416 MTOR NM_004958 2487
## 337843 chr1 11169418 11169418 MTOR NM_004958 2486</code></pre>
<div id="importing-bed12-files-as-input-regions" class="section level2">
<h2>Importing BED12 Files as Input Regions</h2>
<p>For elements and sites written in BED12 files, the prepare_elements_from_BED12 function can directly read the file and adapt it to fulfill the format requirements of the elements and sites parameters of the ActiveDriverWGS function. For more information on the BED12 format, please refer to the UCSC guidelines (<a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1" class="uri">https://genome.ucsc.edu/FAQ/FAQformat.html#format1</a>). In this example, regions are adapted from annotations for protein coding genes from GENCODE.v19 for chromosome 17.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">elements =<span class="st"> </span><span class="kw">prepare_elements_from_BED12</span>(
  <span class="kw">system.file</span>(
    <span class="st">"extdata"</span>, 
    <span class="st">"chr17.coding_regions.bed"</span>, 
    <span class="dt">package =</span> <span class="st">"ActiveDriverWGSR"</span>, 
    <span class="dt">mustWork =</span> <span class="ot">TRUE</span>))</code></pre></div>
<pre><code>## 
##  1189  Rows :: Processing row 100  200  300  400  500  600  700  800  900  1000  1100  
##  Preparing Elements Complete 
##  RM 0 lines</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(elements)</code></pre></div>
<pre><code>##     chr starts  ends                                             id
## 1 chr17   6006  6168 gc19_pc.cds::gencode::DOC2B::ENSG00000272636.1
## 2 chr17  11205 11332 gc19_pc.cds::gencode::DOC2B::ENSG00000272636.1
## 3 chr17  11871 11981 gc19_pc.cds::gencode::DOC2B::ENSG00000272636.1
## 4 chr17  13920 13995 gc19_pc.cds::gencode::DOC2B::ENSG00000272636.1
## 5 chr17  22327 22407 gc19_pc.cds::gencode::DOC2B::ENSG00000272636.1
## 6 chr17  30897 31270 gc19_pc.cds::gencode::DOC2B::ENSG00000272636.1</code></pre>
</div>
</div>
<div id="basic-use" class="section level1">
<h1>Basic Use</h1>
<p>ActiveDriverWGS can be run simply with the mutations file, the elements file and an optional sites file. In this example, mutations are adapted from the Alexandrov et al, 2013 dataset for the CLL patients. Regions are adapted from the cancer gene census and annotations for protein coding genes from GENCODE.v19.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">some_genes =<span class="st"> </span><span class="kw">c</span>(<span class="st">"ATM"</span>, <span class="st">"MYD88"</span>, <span class="st">"NOTCH1"</span>,<span class="st">"SF3B1"</span>, <span class="st">"XPO1"</span>, <span class="st">"SOCS1"</span>, <span class="st">"CNOT3"</span>, <span class="st">"DDX3X"</span>, <span class="st">"KMT2A"</span>, <span class="st">"HIF1A"</span>, <span class="st">"APC"</span>)

results =<span class="st"> </span><span class="kw">ActiveDriverWGS</span>(<span class="dt">mutations =</span> cll_mutations,
                          <span class="dt">elements =</span> cancer_genes[cancer_genes<span class="op">$</span>id <span class="op">%in%</span><span class="st"> </span>some_genes,],
                          <span class="dt">sites =</span> cancer_gene_sites)</code></pre></div>
<pre><code>## 0 remove hypermut, n= 0 ,  0 %
## hypermuted samples:   
## 
## reversing 0 positions
## Tests to do:  11 
## ...........</code></pre>
<div id="parameter-interpretation" class="section level2">
<h2>Parameter Interpretation</h2>
<p>ActiveDriverWGS has several adjustable parameters:</p>
<ol style="list-style-type: decimal">
<li><p><code>window_size</code>: A narrow background window in which mutation rates are assumed to remain constant. We have optimized this parameter on the PCAWG dataset to be 50,000 bps for SNVs and Indels.</p></li>
<li><p><code>filter_hyper_MB</code>: The threshold for the number of simple somatic mutations per megabase above which a sample is considered hypermutated. We define the default to be 30 mutations/megabase according to published literature.</p></li>
<li><p><code>recovery.dir</code>: The directory for writing recovery files for ActiveDriverWGS. If the directory does not exist, it will be created. If none is specified, ActiveDriverWGS will create the directory ActiveDriverWGS_recovery.</p></li>
<li><p><code>mc.cores</code>: The number of cores that the user wishes to allocate to running ActiveDriverWGS. For more information, refer to the R package <code>parallel</code>.</p></li>
</ol>
</div>
</div>
<div id="interpreting-the-results" class="section level1">
<h1>Interpreting the Results</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(results)</code></pre></div>
<pre><code>##       id  pp_element element_muts_obs element_muts_exp element_enriched
## 1    ATM 0.001807467                2                0             TRUE
## 2  MYD88 0.004976353                1                0             TRUE
##     pp_site site_muts_obs site_muts_exp site_enriched result_number
## 1 1.0000000            NA            NA            NA             2
## 2 1.0000000            NA            NA            NA             7
##   fdr_element fdr_site has_site_mutations
## 1  0.01241211        1                   
## 2  0.01241211        1                   
##  [ reached getOption("max.print") -- omitted 4 rows ]</code></pre>
<p>ActiveDriverWGSR will return a results file in a data frame format with the following columns.</p>
<ol style="list-style-type: decimal">
<li><p><code>id</code>: The identifier for the element.</p></li>
<li><p><code>pp_element</code>: The p-value associated with enrichment of mutations in the element.</p></li>
<li><p><code>element_muts_obs</code>: Number of patients with mutations in the element.</p></li>
<li><p><code>element_muts_exp</code>: Number of expected patients with mutations in the element.</p></li>
<li><p><code>element_enriched</code>: Boolean indicating whether an enrichment of mutations is observed.</p></li>
<li><p><code>pp_site</code>: The p-value associated with enrichment of mutations in the sites.</p></li>
<li><p><code>site_muts_obs</code>: Number of patients with mutations in the sites.</p></li>
<li><p><code>site_enriched</code>: Expected number of patients with mutations in the sites.</p></li>
<li><p><code>fdr_element</code>: FDR corrected p-value associated with the element.</p></li>
<li><p><code>fdr_site</code>: FDR corrected p-value associated with the site.</p></li>
</ol>
</div>
<div id="adapting-activedriverwgs-to-local-high-performance-computating-clusters" class="section level1">
<h1>Adapting ActiveDriverWGS to Local High Performance Computating Clusters</h1>
<p>Compute time increases linearly with the number of samples, mutations and regions. Hence, the two main functions integral to ActiveDriverWGS have also been made available in the public domain and can be adapted by the user to their local HPCC.</p>
<div id="format_muts" class="section level3">
<h3>1. format_muts</h3>
<p>This function formats the mutations data frame, removes hyper-mutated samples and removes non-mitochondrial mutations in extrachromosomal regions. It adds an additional column to the mutations data frame that provides the trinucleotide context of the given mutation which will be later used to estimate the mutational distribution across signatures.</p>
</div>
<div id="adwgs_test" class="section level3">
<h3>2. ADWGS_test</h3>
<p>This function calculates the enrichment of mutations for a particular region id. It applies a Poisson generalized linear regression model across mutation signatures to identify enriched regions.</p>
</div>
<div id="example" class="section level2">
<h2>Example</h2>
<p>The following example will demonstrate how to build an ActiveDriverWGS pipeline which can be adapted to HPCCs. The idea is that the list of ids can be split into manageable pieces and run in parallel jobs. Note that the creation of GRanges objects is part of the ActiveDriverWGS wrapper function and must be completed manually by users wishing to create personalized pipelines. Also, note that multiple testing corrections have be recalculated in additional steps.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(GenomicRanges)

<span class="co"># Loading elements &amp; creating a GRanges object</span>
<span class="kw">data</span>(cancer_genes)
gr_element_coords =<span class="st"> </span><span class="kw">GRanges</span>(<span class="dt">seqnames =</span> cancer_genes<span class="op">$</span>chr,
                            <span class="kw">IRanges</span>(<span class="dt">start =</span> cancer_genes<span class="op">$</span>start,
                                    <span class="dt">end =</span> cancer_genes<span class="op">$</span>end),
                            <span class="dt">mcols =</span> cancer_genes<span class="op">$</span>id)

<span class="co"># Loading sites &amp; creating a GRanges object</span>
<span class="kw">data</span>(cancer_gene_sites)
gr_site_coords =<span class="st"> </span><span class="kw">GRanges</span>(<span class="dt">seqnames =</span> cancer_gene_sites<span class="op">$</span>chr,
                         <span class="kw">IRanges</span>(<span class="dt">start =</span> cancer_gene_sites<span class="op">$</span>start,
                                 <span class="dt">end =</span> cancer_gene_sites<span class="op">$</span>end),
                         <span class="dt">mocols =</span> cancer_gene_sites<span class="op">$</span>id)

<span class="co"># Loading mutations, format muts &amp; creating a GRanges object</span>
<span class="kw">data</span>(cll_mutations)

<span class="co"># format_muts</span>
cll_mutations =<span class="st"> </span><span class="kw">format_muts</span>(cll_mutations,
                            <span class="dt">filter_hyper_MB =</span> <span class="dv">30</span>)</code></pre></div>
<pre><code>## 0 remove hypermut, n= 0 ,  0 %
## hypermuted samples:   
## 
## reversing 0 positions</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gr_maf =<span class="st"> </span><span class="kw">GRanges</span>(cll_mutations<span class="op">$</span>chr,
                 <span class="kw">IRanges</span>(<span class="dt">start =</span> cll_mutations<span class="op">$</span>pos1,
                         <span class="dt">end =</span> cll_mutations<span class="op">$</span>pos2),
                 <span class="dt">mcols=</span>cll_mutations[,<span class="kw">c</span>(<span class="st">"patient"</span>, <span class="st">"tag"</span>)])

<span class="co"># Examplifying the TP53 Element</span>
id =<span class="st"> "ATM"</span>

result =<span class="st"> </span><span class="kw">ADWGS_test</span>(<span class="dt">id =</span> id,
                    <span class="dt">gr_element_coords =</span> gr_element_coords,
                    <span class="dt">gr_site_coords =</span> gr_site_coords,
                    <span class="dt">gr_maf =</span> gr_maf,
                    <span class="dt">win_size =</span> <span class="dv">50000</span>)</code></pre></div>
<pre><code>## .</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">result</code></pre></div>
<pre><code>##      id  pp_element element_muts_obs element_muts_exp element_enriched
## 95% ATM 0.001807467                2                0             TRUE
##     pp_site site_muts_obs site_muts_exp site_enriched
## 95%      NA            NA            NA            NA</code></pre>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>



</body></html>