---
title: "ActiveDriverWGSR"
author: "Helen Zhu, Dr. Juri Reimand"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_vignette:
    toc: false
vignette: >
  %\VignetteIndexEntry{ActiveDriverWGSR}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      warning=FALSE, 
                      message=FALSE, 
                      width=500)
options(max.print=35)
```

# Introduction

ActiveDriverWGS is a driver discovery tool for cancer whole genome sequencing data. It analyzes the mutational burden of SNVs and short Indels in functionally defined regions of interest and returns regions which are significantly mutated compared to a background window. For more information, please refer to the ActiveDriverWGS preprint. https://www.biorxiv.org/content/early/2017/12/19/236802

### Please Note 
The genome build for ActiveDriverWGS is hg19. We are working on adding additional options for hg38 and mm10.

# Input Data
ActiveDriverWGSR requires a file for somatic mutations and a file for genomic elements. A third optional file for sites within genomic elements (i.e. post-translational modification sites in protein coding genes) can be used to find genomic elements with enriched site mutations.

```{r input}
library(ActiveDriverWGSR)

data("cll_mutations")
head(cll_mutations)

data("cancer_genes")
head(cancer_genes)

data("cancer_gene_sites")
head(cancer_gene_sites)

```

## Importing BED12 Files as Input Regions
For elements and sites written in BED12 files, the prepare_elements_from_BED12 function can directly read the file and adapt it to fulfill the format requirements of the elements and sites parameters of the ActiveDriverWGS function. For more information on the BED12 format, please refer to the UCSC guidelines (https://genome.ucsc.edu/FAQ/FAQformat.html#format1). In this example, regions are adapted from annotations for protein coding genes from GENCODE.v19 for chromosome 17.

```{r prepare_elements_from_BED12}
elements = prepare_elements_from_BED12(
  system.file(
    "extdata", 
    "chr17.coding_regions.bed", 
    package = "ActiveDriverWGSR", 
    mustWork = TRUE))

head(elements)
```

# Basic Use
ActiveDriverWGS can be run simply with the mutations file, the elements file and an optional sites file. In this example, mutations are adapted from the Alexandrov et al, 2013 dataset for the CLL patients. Regions are adapted from the cancer gene census and annotations for protein coding genes from GENCODE.v19.

```{r ActiveDriverWGS}

some_genes = c("ATM", "MYD88", "NOTCH1","SF3B1", "XPO1", "SOCS1", "CNOT3", "DDX3X", "KMT2A", "HIF1A", "APC")

results = ActiveDriverWGS(mutations = cll_mutations,
                          elements = cancer_genes[cancer_genes$id %in% some_genes,],
                          sites = cancer_gene_sites)

```

## Parameter Interpretation
ActiveDriverWGS has several adjustable parameters:

1) `window_size`: A narrow background window in which mutation rates are assumed to remain constant. We have optimized this parameter on the PCAWG dataset to be 50,000 bps for SNVs and Indels.

2) `filter_hyper_MB`: The threshold for the number of simple somatic mutations per megabase above which a sample is considered hypermutated. We define the default to be 30 mutations/megabase according to published literature.

3) `recovery.dir`: The directory for writing recovery files for ActiveDriverWGS. If the directory does not exist, it will be created. If none is specified, ActiveDriverWGS will create the directory ActiveDriverWGS_recovery.

4) `mc.cores`: The number of cores that the user wishes to allocate to running ActiveDriverWGS. For more information, refer to the R package `parallel`.

# Interpreting the Results

```{r pressure}
head(results)

```

ActiveDriverWGSR will return a results file in a data frame format with the following columns.

1) `id`: The identifier for the element.

2) `pp_element`: The p-value associated with enrichment of mutations in the element.

3) `element_muts_obs`: Number of patients with mutations in the element.

4) `element_muts_exp`: Number of expected patients with mutations in the element.

5) `element_enriched`: Boolean indicating whether an enrichment of mutations is observed.

6) `pp_site`: The p-value associated with enrichment of mutations in the sites. 

7) `site_muts_obs`: Number of patients with mutations in the sites.

8) `site_enriched`: Expected number of patients with mutations in the sites.

9) `fdr_element`: FDR corrected p-value associated with the element.

10) `fdr_site`: FDR corrected p-value associated with the site.


# Adapting ActiveDriverWGS to Local High Performance Computating Clusters
Compute time increases linearly with the number of samples, mutations and regions. Hence, the two main functions integral to ActiveDriverWGS have also been made available in the public domain and can be adapted by the user to their local HPCC. 

### 1. format_muts
This function formats the mutations data frame, removes hyper-mutated samples and removes non-mitochondrial mutations in extrachromosomal regions. It adds an additional column to the mutations data frame that provides the trinucleotide context of the given mutation which will be later used to estimate the mutational distribution across signatures.

### 2. ADWGS_test
This function calculates the enrichment of mutations for a particular region id. It applies a Poisson generalized linear regression model across mutation signatures to identify enriched regions.

## Example
The following example will demonstrate how to build an ActiveDriverWGS pipeline which can be adapted to HPCCs. The idea is that the list of ids can be split into manageable pieces and run in parallel jobs. Note that the creation of GRanges objects is part of the ActiveDriverWGS wrapper function and must be completed manually by users wishing to create personalized pipelines. Also, note that multiple testing corrections have be recalculated in additional steps.

```{r}
library(GenomicRanges)

# Loading elements & creating a GRanges object
data(cancer_genes)
gr_element_coords = GRanges(seqnames = cancer_genes$chr,
                            IRanges(start = cancer_genes$start,
                                    end = cancer_genes$end),
                            mcols = cancer_genes$id)

# Loading sites & creating a GRanges object
data(cancer_gene_sites)
gr_site_coords = GRanges(seqnames = cancer_gene_sites$chr,
                         IRanges(start = cancer_gene_sites$start,
                                 end = cancer_gene_sites$end),
                         mocols = cancer_gene_sites$id)

# Loading mutations, format muts & creating a GRanges object
data(cll_mutations)

# format_muts
cll_mutations = format_muts(cll_mutations,
                            filter_hyper_MB = 30)

gr_maf = GRanges(cll_mutations$chr,
                 IRanges(start = cll_mutations$pos1,
                         end = cll_mutations$pos2),
                 mcols=cll_mutations[,c("patient", "tag")])

# Examplifying the TP53 Element
id = "ATM"

result = ADWGS_test(id = id,
                    gr_element_coords = gr_element_coords,
                    gr_site_coords = gr_site_coords,
                    gr_maf = gr_maf,
                    win_size = 50000)

result
```

